Add architecture design doc (#823)
theforeman/foreman_rh_cloud: ARCHITECTURE.md

# Architecture

> This plugin is intended to connect Foreman instance to Red
> Hat cloud services.

Reviewer: ```suggestion
This plugin is intended to connect a Foreman instance to the Red
Hat Hybrid Cloud Console.
```

Currently it deals with the following aspects:

* Inventory reports
* Recommendations and remediations synchronization
* Cloud side initiated remediations
* Handling of client tools requests from the managed hosts

> Server side is built as a rails engine + react pages that

Reviewer: ```suggestion
Server side is built as a Rails engine + React pages that
```

are used for displaying information that is not related to

> existing Foreman objects.

Reviewer: I am not sure I understand this statement.

In addition to these there are React components that are
integrated into existing foreman pages, like hosts index
and details pages by using the slot and fill mechanism.

The plugin is based on asynchronous communication with the
cloud that is done either automatically once a day or

> initiated manually from the plugin specific pages.

Reviewer: ```suggestion
initiated manually from the plugin's UI.
```

stored in the `lib/{namespace}.rb` file.

The client side is a set of React components that are
following the same files layout:

```
webpack
├── InsightsCloudSync
│   ├── Components
│   │   ├── InternalComponent
│   │   │   └── the-same-structure-as-parent
│   ├── index.js
│   ├── InsightsCloudSyncActions.js
│   ├── InsightsCloudSyncConstants.js
│   ├── InsightsCloudSyncHelpers.js
│   ├── InsightsCloudSync.js
│   ├── InsightsCloudSyncReducers.js
│   ├── InsightsCloudSync.scss
│   ├── InsightsCloudSync.test.js
```

- `{component}Actions.js` contains the functions used by
the component.
- `{component}Constants.js` contains the constants for the
component
- `{component}Helpers.js` contains redux accessors and URI
generator methods
- `{component}.js` main code of the component
- `{component}Reducers.js` contains the reducers that
handle the global redux state.
- `{component}.scss` styling for the component
- `{component}.test.js` tests

## Inventory reports

The idea is to generate a report from Foreman's inventory.
Hosts that will be included are those that have a
subscription attached to them. Additionally we want to
indicate for each host whether it was reported to the cloud

> or not.

Reviewer: Does this mean reported previously or some other method?

Those hosts would be recorded to a separate table, where an
option to remove those records from the cloud would be
available.

## Recommendations and remediations

### Synchronization process

This part is responsible to get recommendations data from
the cloud and present it on Foreman native pages. It
introduces two new objects: **rule** - a set of conditions
that describe a problem and **recommendation** - an
indicator that a rule is applicable to a specific host.

First, the process syncs host records from the cloud,
similar to what happens in the inventory status sync. Then
the process downloads the recommendations and rule
definitions from the cloud into its own tables, adds a
recommendations count column on the hosts list page and
recommendations tab on host details page. Additionally
there is a way to see the whole recommendations table on a
separate page in the UI.

The synchronization is done by the following tasks:
1. Recommendations sync task: hits the cloud API and
creates the recommendations record.
2. Rules sync task: hits the cloud API and creates rule
record that will be available to all recommendations for
the same rule.

> ### Foreman originated remediation

Reviewer: ```suggestion
### Foreman initiated remediation
```

Once the records were synced, there is an option to
remediate selected recommendations by running an ansible
playbook. Once the user selects which recommendations are
selected for remediation, a new Remote Execution Job is

> instanciated. This REX job is based on a specific feature

Reviewer: ```suggestion
instantiated. This REX job is based on a specific feature
```

and a template that comes with it. The template uses a
single helper method that accepts the remediation ids list,
downloads the playbook that will contain the selected
remediations and outputs the resulting playbook. The
downloaded playbook is passed then to the template and fed
into the REX job. From this step it's handled by the REX
plugin.

> ### Cloud originated remediation (Cloud connector)

Reviewer: ```suggestion
### Cloud initiated remediation (Cloud Connector)
```

4. The rhcd instance needs to be registered in the
`Sources` cloud registry as a valid listener.

This configuration process is initiated by the "Configure
cloud connector" button in the UI, and is implemented as a
REX job and a subscriber task. The REX job is initiated
with a specific feature, and executes a playbook from
`foreman-operations` collection. One of the steps in the
task is a request to Foreman server to persist the rhcd id.
The subscriber task waits for this task to finish and then
using the rhcd id from the REX job, it calls the Sources
API to properly subscribe the new rhcd instance.

Once the configuration phase is finished the remediation
request will be able to flow from the cloud, through the
worker and hit a new API on the Foreman server.
The API in turn will examine the request, extract the
playbook id from it and initiate a REX job with that id.
The job also has a specific feature and a template. Similar
to the Foreman initiated use case, the template calls a
dedicated helper method, that hits a cloud API to download
the actual playbook and then passes the playbook to the REX
framework through the template.

> ## Client tools requests forwarding

Reviewer: I'd appreciate a bit more in this section about how the proxying happens, and some pointers to the code architecture around this. Also how authentication is handled.
